Support for skin tones in the picker

rubengees opened the pull request:
This would fix #80.

### Contents of this PR

A popup when clicked on an Emoji to show the available skin colors, similar to WhatsApp. It looks like this:

![device-2017-02-18-001345](https://cloud.githubusercontent.com/assets/8021265/23086824/bbae1268-f56f-11e6-8e9b-b4d690f65dde.png)

#### Code changes

- A new `OnEmojiLongClickedListener`, which is delegated through `Adapters` and `Views`, similar to the existing `OnEmojiClickedListener`. It is used to detect long clicks and show the popup.
- An additional method for finding skin toned Emojis, given an `Emoji` instance, for the `EmojiTree` and `EmojiManager`.
- A new class `EmojiSkinTonePopup`, which is used in the `EmojiPopup` and displays the popup.
- Small cleanups here and there.

rubengees commented on build.gradle:
```diff
@@ -54,6 +54,7 @@ ext {
     supportV4CoreUi = "com.android.support:support-core-ui:$supportLibraryVersion"
     supportV4Compat = "com.android.support:support-compat:$supportLibraryVersion"
     supportRecyclerView = "com.android.support:recyclerview-v7:${supportLibraryVersion}"
+    supportCardView = "com.android.support:cardview-v7:${supportLibraryVersion}"
```
Is it okay to add the `cardview` dependency? I found no other good looking way to add elevation to the new popup, while maintaining backwards compability.

vanniktech replied:
I'd prefer if we would not use it. There must be a way to add elevation to it without CardView.

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java:
```diff
@@ -10,20 +10,27 @@
 import static com.vanniktech.emoji.Utils.checkNotNull;
 
 final class EmojiArrayAdapter extends ArrayAdapter<Emoji> {
+  private static final String EMOJI_NULL_ERROR = "emoji == null";
```
remove this, we'll adjust the PMD / Checkstyle config

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java:
```diff
@@ -55,6 +62,23 @@ public View getView(final int position, final View convertView, @NonNull final V
       }
     });
 
+    if (!EmojiManager.getInstance().findSkinTonedEmojis(checkNotNull(getItem(position), EMOJI_NULL_ERROR)).isEmpty()) {
```
you can extract the variable `checkNotNull(getItem(position), EMOJI_NULL_ERROR)` here

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java:
```diff
@@ -55,6 +62,23 @@ public View getView(final int position, final View convertView, @NonNull final V
+      image.setOnLongClickListener(new View.OnLongClickListener() {
+          @Override
+          public boolean onLongClick(final View v) {
+            if (longListener == null) {
```
invert if statement and remove the else block

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiGridView.java:
```diff
@@ -24,8 +26,11 @@
+                            @Nullable final OnEmojiLongClickedListener onEmojiLongClickedListener,
+                            @NonNull final EmojiCategory category) {
+    emojiArrayAdapter = new EmojiArrayAdapter(getContext(), category.getEmojis(),
+            onEmojiClickedListener, onEmojiLongClickedListener);
```
put it in one line

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiGridView.java:
```diff
@@ -4,8 +4,10 @@
 import android.support.annotation.NonNull;
 import android.support.annotation.Nullable;
 import android.widget.GridView;
+
```
nit: remove

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiArrayAdapter.java:
```diff
@@ -10,20 +10,27 @@
 import com.vanniktech.emoji.emoji.Emoji;
 import com.vanniktech.emoji.listeners.OnEmojiClickedListener;
+import com.vanniktech.emoji.listeners.OnEmojiLongClickedListener;
+
```
nit: remove

rubengees replied:
Android Studio is always trying to change it back, is it really necessary?
Out of curiosity, why does your code style differ so much from the Android Studio defaults? I have hard times getting Android Studio to not reformat 😁

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java:
```diff
@@ -2,10 +2,13 @@
 
 import android.support.annotation.NonNull;
 import android.support.annotation.Nullable;
+
```
nit: remove

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java:
```diff
@@ -2,10 +2,13 @@
 import com.vanniktech.emoji.emoji.Emoji;
 import com.vanniktech.emoji.emoji.EmojiCategory;
 import com.vanniktech.emoji.emoji.EmojiTree;
 
```
nit: remove

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiManager.java:
```diff
@@ -2,10 +2,13 @@
 import com.vanniktech.emoji.emoji.EmojiTree;
 
+import java.util.List;
+
```
nit: remove

rubengees replied:
`checkstyle` complains when removing the space between normal and static imports, so I will keep it as it is.

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiPagerAdapter.java:
```diff
@@ -29,7 +34,8 @@
     } else {
-      newView = new EmojiGridView(pager.getContext()).init(listener, EmojiManager.getInstance().getCategories()[position - 1]);
+      newView = new EmojiGridView(pager.getContext()).init(listener, longListener,
+              EmojiManager.getInstance().getCategories()[position - 1]);
```
can be one line

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiPopup.java:
```diff
@@ -16,9 +16,11 @@
 import android.view.WindowManager;
 import android.view.inputmethod.InputMethodManager;
 import android.widget.PopupWindow;
+
```
nit: remove

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java:
```diff
@@ -0,0 +1,91 @@
+
+import java.util.List;
+
+class EmojiSkinTonePopup {
```
final

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java:
```diff
@@ -0,0 +1,91 @@
+    skinTonedEmojis.add(0, emoji);
+
+    for (final Emoji tonedEmoji : skinTonedEmojis) {
+      final ImageView emojiImage = (ImageView) LayoutInflater.from(clickedImage.getContext())
```
`LayoutInflater.from(clickedImage.getContext())` into variable and before the loop so we don't look it up in a loop

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java:
```diff
@@ -0,0 +1,91 @@
+      final ViewGroup.MarginLayoutParams layoutParams = (ViewGroup.MarginLayoutParams) emojiImage.getLayoutParams();
+      final int margin = (int) Utils.dpToPx(clickedImage.getContext(), MARGIN);
+
+      // Use the same size for Emojis as the picker
```
http://nedbatchelder.com/blog/201401/comments_should_be_sentences.html

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java:
```diff
@@ -0,0 +1,91 @@
+      final ImageView emojiImage = (ImageView) LayoutInflater.from(clickedImage.getContext())
+              .inflate(R.layout.emoji_item, imageContainer, false);
+      final ViewGroup.MarginLayoutParams layoutParams = (ViewGroup.MarginLayoutParams) emojiImage.getLayoutParams();
+      final int margin = (int) Utils.dpToPx(clickedImage.getContext(), MARGIN);
```
cast seems redundant

rubengees replied:
`dpToPx` returns `float`, but we need `int` here.
The code is from [this](http://stackoverflow.com/a/12147550/4279995) StackOverflow answer.

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java:
```diff
@@ -0,0 +1,91 @@
+      imageContainer.addView(emojiImage);
+    }
+
+    content.measure(View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED),
```
static import `makeMeasureSpec`

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/EmojiSkinTonePopup.java:
```diff
@@ -0,0 +1,91 @@
+            View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED));
+
+    popupWindow.setOutsideTouchable(true);
+    popupWindow.setBackgroundDrawable(new BitmapDrawable(clickedImage.getContext().getResources(), (Bitmap) null));
```
What is this line doing?

rubengees replied:
I stole it from [here](https://github.com/vanniktech/Emoji/blob/master/emoji/src/main/java/com/vanniktech/emoji/EmojiPopup.java#L100), I don't know if it is really required though (No visible change with or without).

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/RecentEmojiGridView.java:
```diff
@@ -18,7 +18,7 @@ public RecentEmojiGridView init(@Nullable final OnEmojiClickedListener onEmojiCl
 
     final Collection<Emoji> emojis = recentEmojis.getRecentEmojis();
-    emojiArrayAdapter = new EmojiArrayAdapter(getContext(), emojis.toArray(new Emoji[emojis.size()]), onEmojiClickedListener);
+    emojiArrayAdapter = new EmojiArrayAdapter(getContext(), emojis.toArray(new Emoji[emojis.size()]), onEmojiClickedListener, null);
```
We should also here add the ability to long press and select a different skin tone

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/Utils.java:
```diff
@@ -27,6 +28,10 @@
   }
 
+  public static float dpToPx(final Context context, final float dp) {
+    return dp * context.getResources().getDisplayMetrics().density;
```
do the cast here and return int

rubengees replied:
Ah good point, did not read until here, before I answered above.

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java:
```diff
@@ -5,12 +5,19 @@
 import android.support.annotation.RestrictTo;
 import android.support.v4.util.SparseArrayCompat;
 
```
nit: remove line

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java:
```diff
@@ -43,6 +49,36 @@ public void add(@NonNull final Emoji emoji) {
+    for (int i = 0; i < emoji.getUnicode().length(); i++){
+      current = current.getChild(emoji.getUnicode().charAt(i));
+
+      if (current == null){
```
nit: space between `){`

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java:
```diff
@@ -43,6 +49,36 @@ public void add(@NonNull final Emoji emoji) {
+  public List<Emoji> findSkinTonedEmojis(@NonNull final Emoji emoji){
+    EmojiNode current = root;
+
+    for (int i = 0; i < emoji.getUnicode().length(); i++){
```
extract `emoji.getUnicode()`

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java:
```diff
@@ -43,6 +49,36 @@ public void add(@NonNull final Emoji emoji) {
+
+    current = current.getChild(SKIN_TONE_PART);
+
+    if (current == null) {
```
can you reverse this?

```java
if (current != null) {
  // All of the logic.
  return result;
}

return Collections.emptyList();
```

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java:
```diff
@@ -0,0 +1,11 @@
+
+import com.vanniktech.emoji.emoji.Emoji;
+
+@RestrictTo(RestrictTo.Scope.LIBRARY)
```
same line as interface

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java:
```diff
@@ -0,0 +1,11 @@
+import com.vanniktech.emoji.emoji.Emoji;
+
+@RestrictTo(RestrictTo.Scope.LIBRARY)
+public interface OnEmojiLongClickedListener {
```
make it package private and put it into the root if we do not expose this (to be consistent)

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/listeners/OnEmojiLongClickedListener.java:
```diff
@@ -0,0 +1,11 @@
+
+import android.support.annotation.RestrictTo;
+import android.view.View;
+
```
nit: remove

rubengees commented:
So I investigated elevation further and I see 4 ways:

- Use native elevation: Looks bad on pre-lollipop
- Use a custom `layer-list`: No shadows at top, left and right and the shadow has no gradient

![layer_list](https://cloud.githubusercontent.com/assets/8021265/23094768/48531f30-f5ff-11e6-8dce-60d6fbd5ca24.png)

- Use the `dialog_holo_light_frame` drawable: has too much shadow and does not look really good

![holo](https://cloud.githubusercontent.com/assets/8021265/23094872/fa047ce6-f600-11e6-8b54-83111c33f171.png)

- Use `cardview`: Increases method count and general size

I am still pro-cardview. I looked into it, when using `Proguard` with optimization, it adds only 33 methods and some resources. Thats not too bad IMHO. Plus nearly every App uses `appcompat-v7` and I can imagine many use `cardview` as well. It is also the "official way" for doing this.

What do you think?

I will work on the other things now.

rubengees commented:
Everything else is done!

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java:
```diff
@@ -43,6 +49,47 @@ public void add(@NonNull final Emoji emoji) {
     return result;
   }
 
+  public List<Emoji> findSkinTonedEmojis(@NonNull final Emoji emoji){
```
Can we add tests for this class?

rubengees replied:
Will do once all features are in place~

vanniktech commented on emoji/src/main/java/com/vanniktech/emoji/emoji/EmojiTree.java:
```diff
@@ -43,6 +49,47 @@ public void add(@NonNull final Emoji emoji) {
+    return Collections.emptyList();
+  }
+
+  public Emoji findNonSkinTonedEmoji(@NonNull final Emoji emoji){
```
Can we add tests for this class?

rubengees commented:
- No there is no reason, I will change that.
- Jup, I will try to replicate that.
- I noticed that not all Emojis have the skin tone modifier at the end, which I assumed :/. I'm thinking of a way to handle this.
- Thats quite hard. We would have to look up for every emoji if another skin tone is selected and then find the appropriate one
- Hmm, should be possible.

vanniktech commented:
@rubengees I just saw that Whatsapp displays every skin tone in the recent one so we can leave the behaviour and I'll take it off the list. About the 4th point, if you want I can also do this once this PR is merged.

rubengees commented:
I have spent some time on the skin tone problem and pushed a new approach:

Every `Emoji` now has references to its variants (skin tones, maybe more in the future) and the variants back to the base `Emoji` without modifications. This also greatly simplifies the `EmojiTree` and leaves the heavy work to the `generator`, which was adjusted for this. One drawback is the slightly increased memory consumption (Due to the additional references), but I think it's not too bad.

I also made the popup close on selection.
About the 2. point: I can't find a drawable like that one WhatsApp and others are using. Do you have an idea where to look?

vanniktech commented:
One more thing, I think that the indent level is off as it shows that it has changed.

rubengees commented:
I have renamed the `SquareImageView` to `EmojiImageView` and extended it to draw the indicator rectangle. I have also changed some of the size and margin values to make it look good.
Moreover I have added some tests and cleaned up the code a bit.

Please have a look!

rubengees commented:
Regarding the indentation: The changes you are seeing are from the variants. I though indenting them a bit is a good idea, to make it visible that they are no base emojis, added directly to the tree (Take a closer look at the parentheses: They are passed to the constructor of their respective base emoji).

rubengees commented:
Formatting adjusted!

vanniktech commented:
Thank you once again 👍
